### ISSUE TITLE
generic-worker failure: [mounts] Not able to rename dir

### ISSUE THREAD

**Issue by ahal** [taskcluster/taskcluster#5396]
**Describe the bug**
MacOS builds in mozillavpn recently started failing with this generic-worker error:
https://firefox-ci-tc.services.mozilla.com/tasks/CUSlv4wpQlCPUoyGbmWcBw/runs/0/logs/public/logs/live.log

Afaict, these tasks have not changed recently. Heitor and I have been standing up signing tasks for these builds, but the signing tasks use a different pool, so not sure how that would be related.

**Taskcluster version**
44.11.2

**Comment by ahal** [comment]
Here's the worker running these tasks:
https://firefox-ci-tc.services.mozilla.com/provisioners/releng-hardware/worker-types/mozillavpn-b-1-osx/workers/mdc1/macmini-r8-179?sortBy=resolved&sortDirection=desc

It's possible we got into some kind of bad state somehow...

**Comment by ahal** [comment]
Looks like we've hit this issue before:
https://bugzilla.mozilla.org/show_bug.cgi?id=1716287

Though I'm not sure that we've ever diagnosed *why* it happens. Possibly there's a bug in `generic-worker` that can sometimes leave workers in this state? Maybe @davehouse has some context from the first time around.

**Comment by ahal** [comment]
This happened again in the VPN repo:
https://mozilla-hub.atlassian.net/browse/RELOPS-148

Though this time, it's in the L3 pool. I think we need to figure out why this keeps happening before we can rely on these builds for releases :/

**Comment by ahal** [comment]
This happened for a 3rd time in 12 days:
https://firefox-ci-tc.services.mozilla.com/provisioners/releng-hardware/worker-types/mozillavpn-b-1-osx/workers/mdc1/macmini-r8-179?sortBy=resolved&sortDirection=desc

At this point I think these pools are more or less unusable as we can't keep asking relops to fix them whenever it keeps happening. @petemoore any chance we can bump the priority of this issue? It means we can't switch to Taskcluster for the VPN Mac builds (which we have a lot of work blocked on).

**Comment by petemoore** [comment]
I can look into this tomorrow.

**Comment by ahal** [comment]
Thanks, appreciate it!

**Comment by ahal** [comment]
Awesome, thanks for the patch @petemoore !

Now I can't remember.. will this rollout with the regular FirefoxCI upgrades? Or is there a separate process for the workers? I feel like it's the latter but wanted to check.

**Comment by petemoore** [comment]
> Awesome, thanks for the patch @petemoore !
> 
> Now I can't remember.. will this rollout with the regular FirefoxCI upgrades? Or is there a separate process for the workers? I feel like it's the latter but wanted to check.

Yes, that's right. Whoever manages the worker pools that are affected will need to deploy new workers, once the new release is made.

**Comment by ahal** [comment]
Thanks! I'll ping relops to see if they can test this out. The two pools that seem to be affected here have a grand total of 3 workers, so it might actually be a good enough fix for now :p

**Comment by ahal** [comment]
Hmm, this may not be fixed. Looks like it's happening again:
https://firefox-ci-tc.services.mozilla.com/provisioners/releng-hardware/worker-types/mozillavpn-b-1-osx/workers/mdc1/macmini-r8-179?sortBy=resolved&sortDirection=desc

I'll have to confirm with @davehouse that the `generic-worker` binary he copied onto there didn't somehow get re-puppetized or something.

**Comment by petemoore** [comment]
> Hmm, this may not be fixed. Looks like it's happening again: https://firefox-ci-tc.services.mozilla.com/provisioners/releng-hardware/worker-types/mozillavpn-b-1-osx/workers/mdc1/macmini-r8-179?sortBy=resolved&sortDirection=desc
> 
> I'll have to confirm with @davehouse that the `generic-worker` binary he copied onto there didn't somehow get re-puppetized or something.

It looks like the correct version of generic-worker - so indeed it looks like this fix didn't work! 👎 😭

**Comment by petemoore** [comment]
That was my bad - I suggested trying it before we had a release, and didn't mention it to you, sorry Boris! I followed the "source" link above, and it looks like it includes my "fix".

**Comment by petemoore** [comment]
@davehouse Are the worker logs in papertrail? I'd like to take a look at them to see if I can work out what is going on.

**Comment by ahal** [comment]
I asked Michelle to dig up the papertrail for this worker and she says that it appears to not be connected :/

**Comment by petemoore** [comment]
The worker logs should still be on the machine though, is it possible to send them? I think under /var/log/generic-worker iirc

**Comment by petemoore** [comment]
OK, I see the problem (at least part of it).

The worker isn't rebooting between tasks, so it isn't logging in as a new task user each time. When I look through the recent tasks, they are _all_ using task_163336061065616 user. That's not how the multiuser engine is intended to be used.

At this point, I don't know what caused the first exception, but subsequent exceptions are caused by the machine not rebooting after the task resolves.

There is an option in generic-worker to shutdown on internal error, which is probably disabled, since this isn't a cloud worker but a long lived physical machine or VM. However, there currently isn't an option to reboot on internal error. Perhaps we should add this. In the meantime it should be sufficient to check the exit code of the worker, and reboot if it is 69, which is the code for an internal error / worker panic.

It still doesn't tell us why the first error occurred, but at least it tells us why after one exception occurs, that the worker stays in the exception state. I'll see if I can trace back to the initial exception...

**Comment by petemoore** [comment]
> The worker isn't rebooting between tasks, so it isn't logging in as a new task user each time. When I look through the recent tasks, they are _all_ using task_163336061065616 user.

The integer part of the task user name is the unix timestamp of the time the user was created. This timestamp maps to 4 October 2021, which means it was created over 7 months ago. That suggests this worker has been using the same task user account for all this time, or that something is resetting machine state back to something from 7 months ago, or that the clock is 7 months off.

Looking at [another worker](https://firefox-ci-tc.services.mozilla.com/provisioners/releng-hardware/worker-types/mozillavpn-b-3-osx/workers/mdc1/macmini-r8-181?sortBy=started&sortDirection=desc) in the same pool, I see green jobs, up until the two most recent. The [most recent successful task](https://firefox-ci-tc.services.mozilla.com/tasks/LyxiNhOPSaue6-m_uT9elA/runs/0/logs/public/logs/live.log) from 21 hours ago used task user `task_163397093166975` which maps to 11 October 2021.

The task user name should have a timestamp from the time it was created, which should be around the time it started the previous task. Something is definitely up here.

**Comment by petemoore** [comment]
So indeed, looking through history of that worker too, it never reboots, even for successful tasks.
That's the reason for all the problems.

**Comment by ahal** [comment]
Thanks for the investigation! Sounds like this is something that will need to be fixed on the relops side. This is happening frequently enough that it might still be worth investigating the root cause.. but it will definitely be a much smaller priority if we can get a clean slate between each task so the failure isn't recurring.

**Comment by ahal** [comment]
Is it normally `generic-worker` that determines the user?

@petemoore would you mind re-opening this issue? I don't have permission.

**Comment by ahal** [comment]
@petemoore I think it was in the community meeting you mentioned something about these workers using `simple` engine or something like that, which was only supposed to be for development. Could that be why the user never changes?

**Comment by ahal** [comment]
Looks like we're able to work around this by disabling caches in the affected tasks. We should still figure out what's going on here and fix it, but at least it's not blocking VPN automation anymore.

**Comment by ahal** [comment]
I'll leave that up to @petemoore. There may be an argument to be made that it shouldn't be possible for workers to be misconfigured in this way, and / or that `generic-worker` should detect and warn when they are.

**Comment by petemoore** [comment]
Yeah, that is fair. I think generic-worker could reasonably mark OS users as having been "used", and not allow them to be used more than once. I'm not sure if it is best to have a separate issue for that, or use this one. Probably would be cleanest to open that as a new issue, and close here. I'll do that.
